seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java

```diff
@@ -40,15 +43,34 @@
 import java.util.Optional;
 import java.util.ServiceLoader;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
 import java.util.stream.Collectors;
 
 public abstract class AbstractPluginDiscovery<T> implements PluginDiscovery<T> {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPluginDiscovery.class);
     private final Path pluginDir;
 
+    /**
+     * Add jar url to classloader. The different engine should have different logic to add url into
+     * their own classloader
+     */
+    private BiConsumer<ClassLoader, URL> addURLToClassLoader = (classLoader, url) -> {
+        if (classLoader instanceof URLClassLoader) {
+            ReflectionUtils.invoke(classLoader, "addURL", url);
+        } else {
+            throw new UnsupportedOperationException("can't support custom load jar");
+        }
+    };
+
     protected final ConcurrentHashMap<PluginIdentifier, Optional<URL>> pluginJarPath =
-        new ConcurrentHashMap<>(Common.COLLECTION_SIZE);
+            new ConcurrentHashMap<>(Common.COLLECTION_SIZE);
+
+    public AbstractPluginDiscovery(String pluginSubDir, BiConsumer<ClassLoader, URL> addURLToClassloader) {
```

Comment:
It's better to move the addURL logic to this class, and the other class doesn't need to submit this parameter to tell the AbstractPluginDiscovery how to add URL, since we don't expose the classloader to them, so they don't know which classloader we will use.
```suggestion
    public AbstractPluginDiscovery(String pluginSubDir) {
```

Reply:
This BiConsumer mean engine use their classloader to add url into itself. So caller should tell `AbstractPluginDiscovery ` how to add url into  their classloader. The `AbstractPluginDiscovery` is without engine. So it can't decide how to add url into classloader( but also have default implement).

Reply:
But the classloader is created by `AbstractPluginDiscovery `, the sub engine discovery doesn't know classloader. 

If you expose these in constructor, the caller need to understand what this function means, then he needs to know the whole logic how we load the plugin.

Reply:
Anyway, I approve this first, since we need to fix #2212